src/Observable.ts

```diff
@@ -1,9 +1,8 @@
 export type TEventCallback<T = any> = (options: T) => any;
 
-type EventRegistryObject<
-  K extends string | number | symbol = string,
-  E = any
-> = Record<K, TEventCallback<E>>;
+type EventRegistryObject<E = any> = {
```

Comment:
should E extend `Record<string, any>`?
Then `K` should always be string and no symbol etc.

Reply:
> should E extend Record<string, any>?

I tried it but it doesn't seem to help to restrict the key type to `string`. In turn it would force `EventSpec` to be a `Record` it self, which is okay, I just didn't do it to avoid in turn affecting maybe other types as well. But if you're okay with it, I can try.

Reply:
Okay I've tried it and it still gives this error

<img width="1058" alt="Screenshot 2023-07-29 at 13 09 09" src="https://github.com/fabricjs/fabric.js/assets/10067273/df886775-cd3f-48db-b2df-50c2c2ee55ba">

Using `Object.entries` fixes the issue, probably `for (key in obj)` is hardcoded to have type `string | number | symbol` because you're looping over all the keys, even non-own properties.

However, `extends Record<string, any>` brought to `CommonMethods<CanvasEvents>` giving error because `CanvasEvents` cannot be assigned to `Record<string, any>`. It works only with `type`, not `interface`. But we need `interface` for declaration merging.

See:
- https://github.com/microsoft/TypeScript/issues/15300
- https://www.typescriptlang.org/play?#code/C4TwDgpgBAogbhAdsAShA5gSwM7AE4gA8MUEAHsEgCbZRoDGA9nlYbnpougDRSICuAWwBGEPAD5xUALxQA3lADaAaSicoAawghGAM1gBdAFx8hovFAC+AKGuhIsBMgDKkejLoQmLNvk49TETFxa05KPF0AQ3poeCRgVy8AJnlrKHSoSJMk6xs7cGgAQQ845DQsdiJShLcQ+2gAIRKnVAwcfCqWxPok8SA

Reply:
Does [this](https://www.typescriptlang.org/play?#code/C4TwDgpgBAogbhAdsAShA5gSwM7AE4gA8MUEAHsEgCbZRoDGA9nlYbnpougDRSICuAWwBGEPAD5xUALxQA3lADaAaSicoAawghGAM1gBdAFx8hovFAC+AKGuhIsBMgDKkejLoQmLNvk49TETFxa05KPF0AQ3poeCRgVy8AJlIKalo4lzd5ayg8qEiTJOsbO3BoAEEPTNQMHHwiGsT6EPtoACFqp1qsdkbu5qTxIA) solve the issue?

I didn't fully understand the interface limitation. Didn't read it thoroughly.
This is when I say ohh TS... why???

Reply:
I am ok using Object.entries
Very annoying the for in loop wrong stqtic types

Reply:
Another question about declaration merging
Why not pass an augmented type to the class? That was my intention we I wrote it. I didn't think of declaratiin mergign because I don't use it so much.
I am used to generics

Reply:
So if i understand correctly using types is better for maintainers because TS is simpler to use and interfaces for devs/consumers for declarstiin merging but that takes a toll from maintaining?

Reply:
> I am ok using Object.entries Very annoying the for in loop wrong static types Just for discussion, do you know of a perf diff between them?

Or we just `@ts-expect-error`

Reply:
> Does [this](https://www.typescriptlang.org/play?#code/C4TwDgpgBAogbhAdsAShA5gSwM7AE4gA8MUEAHsEgCbZRoDGA9nlYbnpougDRSICuAWwBGEPAD5xUALxQA3lADaAaSicoAawghGAM1gBdAFx8hovFAC+AKGuhIsBMgDKkejLoQmLNvk49TETFxa05KPF0AQ3poeCRgVy8AJlIKalo4lzd5ayg8qEiTJOsbO3BoAEEPTNQMHHwiGsT6EPtoACFqp1qsdkbu5qTxIA) solve the issue?

Not really, because at that point you can do `const value: EventSpec2 = { a: 2, b: 3 }`, e.g. add extra properties without error. And that's exactly why TS gives the error in my example, because technically you're saying that something with "well-defined properties" is assignable to a type where "any string index can be used", which is unsafe.

> Just for discussion, do you know of a perf diff between them?

![Screenshot 2023-07-30 at 12 43 07](https://github.com/fabricjs/fabric.js/assets/10067273/866b9f17-113a-4378-a27d-ec9e9aef8f84)

There is definitely an overhead in using functions, but not significant. Also not relevant for an object of event handlers, whose number of entries is only a few and the code is not in a critical hot path (rendering).

> Why not pass an augmented type to the class? That was my intention when I wrote it. I didn't think of declaration merging because I don't use it so much. I am used to generics

Declaration merging is different than generics. Generic is for a specific instance, a declaration merging extends an existing interface wherever it's used. It can also significantly harder for the lib to maintain good generics. For instance the following would need to be fixed as the `EventSpec` generic is currently not used:

<img width="1095" alt="Screenshot 2023-07-30 at 12 45 29" src="https://github.com/fabricjs/fabric.js/assets/10067273/1d3f6f52-b055-41e8-8e50-434168993396">

But if you fix everything and make it work, it's definitely a doable alternative. At that point not sure if there are still other use cases where declaration merging for canvas/object events might be helpful.